File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/GeneralStyleSettingsSerializer.kt
package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.GeneralStyleSettings
import com.usercentrics.sdk.flutter.api.FlutterActivityProvider
import com.usercentrics.sdk.flutter.api.FlutterAssetsProvider

internal fun Any?.deserializeGeneralStyleSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): GeneralStyleSettings? {
    if (this !is Map<*, *>) {
        return null
    }
    return GeneralStyleSettings(
        textColor = (this["textColor"] as? String)?.deserializeColor(),
        layerBackgroundColor = (this["layerBackgroundColor"] as? String)?.deserializeColor(),
        layerBackgroundSecondaryColor = (this["layerBackgroundSecondaryColor"] as? String)?.deserializeColor(),
        linkColor = (this["linkColor"] as? String)?.deserializeColor(),
        tabColor = (this["tabColor"] as? String)?.deserializeColor(),
        bordersColor = (this["bordersColor"] as? String)?.deserializeColor(),
        toggleStyleSettings = this["toggleStyleSettings"]?.deserializeToggleStyleSettings(),
        logo = this["logo"].deserializeImage(assetsProvider, activityProvider),
        font = this["font"].deserializeBannerFont(assetsProvider, activityProvider),
        links = this["links"].deserializeLegalLinksSettings(),
        disableSystemBackButton = this["disableSystemBackButton"] as? Boolean,
    )
}

File: lib/src/internal/serializer/general_style_settings_serializer.dart
import 'package:usercentrics_sdk/src/internal/serializer/color_serializer.dart';
import 'package:usercentrics_sdk/src/internal/serializer/legal_links_settings_serializer.dart';
import 'package:usercentrics_sdk/src/internal/serializer/toggle_style_settings_serializer.dart';
import 'package:usercentrics_sdk/src/model/font.dart';
import 'package:usercentrics_sdk/src/model/general_style_settings.dart';

class GeneralStyleSettingsSerializer {
  static dynamic serialize(GeneralStyleSettings? value) => value == null
      ? null
      : {
          'textColor': ColorSerializer.serialize(value.textColor),
          'layerBackgroundColor':
              ColorSerializer.serialize(value.layerBackgroundColor),
          'layerBackgroundSecondaryColor':
              ColorSerializer.serialize(value.layerBackgroundSecondaryColor),
          'linkColor': ColorSerializer.serialize(value.linkColor),
          'tabColor': ColorSerializer.serialize(value.tabColor),
          'bordersColor': ColorSerializer.serialize(value.bordersColor),
          'toggleStyleSettings': ToggleStyleSettingsSerializer.serialize(
              value.toggleStyleSettings),
          'font': _serializeFont(value.font),
          'logo': value.logo?.assetPath,
          'links': LegalLinksSettingsSerializer.serialize(value.links),
          'disableSystemBackButton' : value.disableSystemBackButton,
        };

  static dynamic _serializeFont(BannerFont? customFont) {
    return customFont == null
        ? null
        : {
            'regularFontAssetPath': customFont.regularFontAssetPath,
            'boldFontAssetPath': customFont.boldFontAssetPath,
            'fontSize': customFont.fontSize,
          };
  }
}
